### PR TITLE
Model, Routes, Controllers for Messaging Feature

### LINKED ISSUE #3: BE: Messaging
Might require multiple models (conversation, messages)
Routes:
- Create new conversation
- Send message
- Get all conversations
- Get all messages for a conversation

### PR DESCRIPTION
**Pull request by EdwardOlmos**
### What this PR does (required):
- Resolves #3
- Adds Conversation and Message Models
- Adds Controllers and Routes with validation for
  - POST /conversations - Create a new conversation
  - POST /messages/:conversationId - Create a new message
  - GET /conversations - Return all user's conversations
  - GET /conversations/:conversationId/messages - Return messages from conversation

### Any information needed to test this feature (required):
Use Postman to perform these calls. All routes are protected, ensure cookie with a valid token is being used to perform the following calls. Have 1 Dog Sitter and 1 Dog Owner account.
  - POST /conversations - Create a new conversation
    - JSON Body: `{ dogOwnerId: "dogOwnerId", dogSitterId: "dogSitterId" }`
      - Response: Status Code: 201 - `{ "success": "Successfully created conversation" }`
  - POST /messages/:conversationId - Create a new message
    - JSON Body: `{ dogOwnerId: "dogOwnerId", dogSitterId: "dogSitterId" }`
      - Response: Status Code: 201 - `{ "success": "Successfully created message" }`
  - GET /conversations - Return all user's conversations
    - Response: Status Code: 200 - `{ "success" : { "conversations": [ Conversation Objects ] } }`
  - GET /conversations/:conversationId/messages - Return messages from conversation
    - Response: Status Code: 200 - `{ "success" : { "messages": [ Message Objects ] } }`

### CODE REVIEW

**Review by ezeilo-su — COMMENTED** [review]
Nice job on this one.

**Comment by ezeilo-su** [comment]
Thanks for making requested changes : +1:

File: server/controllers/conversation.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+const asyncHandler = require("express-async-handler");
+
+// @route POST /conversations
+// @desc Create new conversation
+// @access Public
+exports.createConversation = asyncHandler(async (req, res, next) => {
+  const { dogOwnerId, dogSitterId } = req.body;
+
+  try {
+    const conversation = await Conversation.create({
```
Comment (rajivtitus) [inline server/controllers/conversation.js:11]: Let's check if a conversation between two users exists before we create a new one :)

File: server/controllers/conversation.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
+const Conversation = require("../models/Conversation");
+const asyncHandler = require("express-async-handler");
+
+// @route POST /conversations
+// @desc Create new conversation
+// @access Public
+exports.createConversation = asyncHandler(async (req, res, next) => {
+  const { dogOwnerId, dogSitterId } = req.body;
+
+  try {
```
Comment (rajivtitus) [inline server/controllers/conversation.js:10]: We don't need a try/catch block as the `asyncHandler` handles promise rejections and forwards them to the error middleware

Ditto throughout!

File: server/controllers/conversation.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    const conversation = await Conversation.create({
+      dogOwnerId,
+      dogSitterId
+    });
+    if (conversation)
+      res.status(201).json({ success: "Successfully created conversation" });
+    else
```
Comment (rajivtitus) [inline server/controllers/conversation.js:17]: We also don't need this `else` clause - if the Conversation document creation has failed, the rejection will be handled by the catch block

Ditto where applicable

File: server/controllers/conversation.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+// @desc Get all conversations
+// @access Public
+exports.getAllConversations = asyncHandler(async (req, res, next) => {
+  const userId = req.user.id;
+
+  const allConversations = await Conversation.find({
+    $or: [{ "petSitter": userId }, { "petOwner": userId }]
+  }).populate("recentMessage");
+
+  if (allConversations) {
```
Comment (rajivtitus) [inline server/controllers/conversation.js:34]:
```
    res.status(200).json({
      success: {
        conversations: allConversations
      }
    });
```
We don't need an if/else clause here as the `find()` query returns an empty list when no matching records are found (an empty list is a truthy value). Ideally we want to forward this empty list to the FE to have a consistent response format :)

File: server/controllers/conversation.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  }
+});
+
+// @route GET /conversations/:conversationId/messages
+// @desc Get all messages from a conversation
+// @access Public
+exports.getMessages = asyncHandler(async (req, res, next) => {
+  const conversationId = req.params.conversationId;
+
+  try {
```
Comment (rajivtitus) [inline server/controllers/conversation.js:51]: Same as comment above about the usage of try/catch

File: server/models/Message.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  },
+  userId: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: 'User',
+    required: true,
+  },
+  message: {
+    type: String,
+    required: true,
+  }
```
Comment (rajivtitus) [inline server/models/Message.js:17]: Additionally, we can also add a `read` status too messages 😄

**Review by rajivtitus — APPROVED** [review]
Excellent work! Great use of mongoose and validation middleware 🦖 
Just some minor comments to address before merging :)

**Pull request merged**